feature_engine/encoding/similarity_encoder.py

```diff
@@ -0,0 +1,298 @@
+from feature_engine.encoding.base_encoder import (
+    CategoricalInitMixin,
+    CategoricalMethodsMixin,
+)
+
+
+def _gpm_fast(x1: str, x2: str) -> float:
+    return difflib.SequenceMatcher(None, x1, x2).quick_ratio()
+
+
+_gpm_fast_vec = np.vectorize(_gpm_fast)
+
+
+@Substitution(
+    ignore_format=_ignore_format_docstring,
+    variables=_variables_categorical_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class StringSimilarityEncoder(CategoricalInitMixin, CategoricalMethodsMixin):
+    """
+    The StringSimilarityEncoder() replaces categorical variables by a set of
+    float variables representing similarity between unique categories in the variable.
+    This new variables will have values in range between 0 and 1, where 0 is the least
+    similar and 1 is the exact match.
+    This encoding is an alternative to OneHotEncoder in the case of poorly
+    defined categorical variables.
+
+    The encoder will create k variables, where k is the number of unique categories.
+
+    The encoder has the additional option to generate similarity variables only for the
+    most popular categories, that is, the categories that are shared by the
+    majority of the observations in the dataset. This behaviour can be specified with
+    the parameter `top_categories`.
+
+    The encoder has the option to specify the behaviour when NaN is present in the
+    variable, see parameter `handle_missing`.
+
+    The encoder will encode only categorical variables by default (type 'object' or
+    'categorical'). You can pass a list of variables to encode. Alternatively, the
+    encoder will find and encode all categorical variables (type 'object' or
+    'categorical').
+
+    With `ignore_format=True` you have the option to encode numerical variables as well.
+    The procedure is identical, you can either enter the list of variables to encode, or
+    the transformer will automatically select all variables.
+
+    The encoder first finds the categories to be encoded for each variable (fit). The
+    encoder then creates one dummy variable per category for each variable
+    (transform).
+
+    Parameters
+    ----------
+    top_categories: int, default=None
+        If None, dummy variables will be created for each unique category of the
+        variable. Alternatively, we can indicate in the number of most frequent
+        categories to encode. In this case, similarity variables will be created
+        only for those popular categories and the rest will be ignored.
+
+    handle_missing : str, default='impute'
+        Action to perform when NaN is seen.
+            'error' - raise an error;
+            'impute' - impute NaN with an empty string;
+            'ignore' - ignore NaN and leave them in resulting columns.
+
+    {variables}
+
+    {ignore_format}
+
```

Comment:
what would this encoder do for unseen categories? should we add parameter unseen and some logic?

Reply:
I think it should be in the notes